### ISSUE TITLE
[MNG-6276] Support Reproducible Builds (https://reproducible-builds.org)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7584]
**[Paolo Sacconier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=psacc)** opened **[MNG-6276](https://issues.apache.org/jira/browse/MNG-6276?redirect=false)** and commented

A venerable build system like Maven should support full build reproducibilty (i.e. producing bit a bit identical binaries from the same source).

As initiatives like https://reproducible-builds.org gain traction and the news of the recent Debian policy change to mandate this build behavior (see https://reproducible.alioth.debian.org/blog/posts/121/), this seems a feature that needs to be considered for inclusion into Maven core & core plugins.

There is an independent ongoing effort to support this feature and the author stated that he has found interest from maven project to integrate his work: https://github.com/Zlika/reproducible-build-maven-plugin/issues/6#issuecomment-325005883

I hope this issue helps kickstart the effort.

---

**Issue Links:**
- [ARCHETYPE-590](https://issues.apache.org/jira/browse/ARCHETYPE-590) support Reproducible Builds for archetype:jar
 (_**"contains"**_)
- [MACR-53](https://issues.apache.org/jira/browse/MACR-53) Reproducible Builds: make entries in output jar files reproducible (order + timestamp)
 (_**"contains"**_)
- [MASSEMBLY-921](https://issues.apache.org/jira/browse/MASSEMBLY-921) Reproducible Builds: make entries in output archive reproducible (order + timestamp)
 (_**"contains"**_)
- [MEAR-280](https://issues.apache.org/jira/browse/MEAR-280) Reproducible Builds: make entries in output ear files reproducible (order + timestamp)
 (_**"contains"**_)
- [MEJB-128](https://issues.apache.org/jira/browse/MEJB-128) Reproducible Builds: make entries in output jar files reproducible (order + timestamp)
 (_**"contains"**_)
- [MJAR-263](https://issues.apache.org/jira/browse/MJAR-263) Reproducible Builds: make entries in output jar files reproducible (order + timestamp)
 (_**"contains"**_)
- [MJAVADOC-627](https://issues.apache.org/jira/browse/MJAVADOC-627) Reproducible Builds: make entries in output jar files reproducible (order + timestamp)
 (_**"contains"**_)
- [MRAR-86](https://issues.apache.org/jira/browse/MRAR-86) Reproducible Builds: make entries in output rar files reproducible (order + timestamp)
 (_**"contains"**_)
- [MSHADE-347](https://issues.apache.org/jira/browse/MSHADE-347) Reproducible Builds: make entries in output jar files reproducible
 (_**"contains"**_)
- [MSITE-851](https://issues.apache.org/jira/browse/MSITE-851) Reproducible Builds: make entries in output jar files reproducible (order + timestamp)
 (_**"contains"**_)
- [MSOURCES-120](https://issues.apache.org/jira/browse/MSOURCES-120) Reproducible Builds: make entries in output jar files reproducible (order + timestamp)
 (_**"contains"**_)
- [MWAR-432](https://issues.apache.org/jira/browse/MWAR-432) Reproducible Builds: make entries in output jar files reproducible (order + timestamp)
 (_**"contains"**_)
- [MJLINK-75](https://issues.apache.org/jira/browse/MJLINK-75) Reproducibility of ZIP artifacts
 (_**"contains"**_)
- [MRRESOURCES-114](https://issues.apache.org/jira/browse/MRRESOURCES-114) make projectTimespan reproducible: use outputTimestamp instead of current date
 (_**"contains"**_)
- [MPLUGIN-326](https://issues.apache.org/jira/browse/MPLUGIN-326) Remove timestamp in generated descriptor
 (_**"contains"**_)
- [MSHARED-494](https://issues.apache.org/jira/browse/MSHARED-494) Impossible to generate a reproducible build due to timestamp in pom.properties

- [MSHARED-787](https://issues.apache.org/jira/browse/MSHARED-787) Add optional buildEnvironment information to the manifest

- [MSHARED-362](https://issues.apache.org/jira/browse/MSHARED-362) Support removing default manifest entries with Maven Archiver

- [MSHARED-511](https://issues.apache.org/jira/browse/MSHARED-511) Using structures to keep insertion order for MANIFEST.MF

- [MSHARED-661](https://issues.apache.org/jira/browse/MSHARED-661) Remove manifest entry "Built-By: \<username>" for reproducible builds

- [MPLUGIN-261](https://issues.apache.org/jira/browse/MPLUGIN-261) sort goals in generated plugin.xml

- [COMPRESS-485](https://issues.apache.org/jira/browse/COMPRESS-485) Reproducible Builds: keep entries order when gathering ScatterZipOutputStream content in ParallelScatterZipCreator

- [MSHARED-796](https://issues.apache.org/jira/browse/MSHARED-796) use java.specification.version instead of java.version in Build-Jdk manifest entry

- [MSHARED-797](https://issues.apache.org/jira/browse/MSHARED-797) Move current Build-Jdk manifest entry to Build-Jdk-Spec

- [MNG-6789](https://issues.apache.org/jira/browse/MNG-6789) Make Maven distribution build Reproducible
 (_**"is required by"**_)
- [MARTIFACT-24](https://issues.apache.org/jira/browse/MARTIFACT-24) add artifact:check-buildplan goal to check that plugins versions do not have known reproducibility issues
 (_**"is superceded by"**_)
- [MNG-8258](https://issues.apache.org/jira/browse/MNG-8258) activate Reproducible Builds by default
 (_**"is depended upon by"**_)

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=74682318)
- [GitHub Pull Request #77
](https://github.com/apache/commons-compress/pull/77)
- [Add ability to NOT write timestamps for zip file entries
](https://github.com/codehaus-plexus/plexus-archiver/issues/48)

9 votes, 17 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I think `idempotent` is a better term, because is clearly states that its result should be unchanged when executed multiple times. And maybe this is can be handled the same way as threadsafety: introduce an annotation which states that the plugin in idempotent aware.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

thank you [Zlika](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Zlika) for the followup: let's continue

we need to find a property name that everybody will agree upon: "reproducible" does not gain momentum yet, nor "idempotent", nor "deterministic"
Re-reading https://reproducible-builds.org/, which seems a good starting point, what about `verifiable`?
To me, finding an agreed property name is the only requirement to fix MSHARED-661, which one of the easiest part to code, then a good concrete first change to do.

To me, finding the right term is not just a detail, but a question of determining the right objective: looking at MSHARED-661, by removing timestamps, the build can be deterministic and idempotent on my personal machine, but our requirement is also that _someone else_ with a "decently near" configuration will get the bit-for-bit same result (then removing username avoids some stupid constraints on build environment configuration)

A general question: is there some writing somewhere on what are the issues in a basic java build? (by "basic" I mean that no advanced build tool like Maven and plugins adds more variable parts)
The first strong issue I see for example in basic builds is _timestamps for files in jars/wars/zips_
Is there something on the precise JDK version used? or compiler? If I build with JDK 8 with target 6, do I get the same .class than with JDK 6? If I build with OpenJDK javac or IBM JDK javac or Eclipse jdt compiler or jikes, do I get the same result as with Oracle JDK javac?

Notice I just added a new entry in https://cwiki.apache.org/confluence/display/MAVEN/Proposals to track this proposal: I'll add a dedicated Wiki page to gather requirements, which will probably be useful on a long term documentation purpose in addition to our discussion in this Jira issue...

**Comment by jira-importer** [comment]
**[Zlika](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zlika)** commented

I agree that naming is very important. I think "Verifiable" is a good name. However, "reproducibility" is clearly used by everyone, even people from other JIRA tickets I linked here. Maven is all about "convention over configuration", so I think we should use the de facto "convention".

Apart from ZIP timestamps, there is another important source of non-reproducibility few people think about: the order of the files in the ZIP, which depends on the file system of the computer.

I didn't do an exhaustive test, but compiling one of my projects with different versions of OpenJDK 8 led to the same class files. I would not be surprised however that an Eclipse or IBM compiler give slightly different results.

**Comment by jira-importer** [comment]
**[Daniel Wegener](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dwegener)** commented

I wrote about this topic last year: https://blog.holisticon.de/2016/10/reproducible-builds-in-java/

TL;DR:
- given the same environment, javac's classfile output is stable
- jar's uses zip which contains entry-timstamps which we must set to a fixed value (or a somewhat non-arbitrary default)
- maven-archiver-plugin may parallelize the packaging which may results in arbitrary entry order in the packaged artifact
- the archiver plugin iterates the files which are to be packaged in file-system order which may differ across plaforms
- I have not tested how the equivalent "standard tooling", the jar command line tool (http://docs.oracle.com/javase/8/docs/technotes/tools/unix/jar.html) behaves

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Your blog post contains at least one error:

> Additionally some older versions of Maven (see MSHARED-494) place the build timestamp into the generated pom.properties file.

This is not correct. The timestamp is written by Sun's `Properties#store()` method, not us.

**Comment by jira-importer** [comment]
**[Daniel Wegener](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dwegener)** commented

Agreed, the formulation is a bit misleading and should have been "newer versions of maven archiver remove the build timestamp generated by suns Properties#store()" ;)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy), it this completed with Maven 3.7.0?

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

in fact, implementation fully happens at plugins level, it is finally 100% independent from Maven version...
I used this existing issue to link issues from our plugins and components that have been updated to produce Reproducible Builds
as noted in the WIki https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=74682318 , there are still a few plugins we maintain to be released

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

Reproducible Builds is now considered done:
- user documentation: https://maven.apache.org/guides/mini/guide-reproducible-builds.html
- artifact:check-buildplan to track plugins with issues: https://maven.apache.org/plugins/maven-artifact-plugin/plugin-issues.html

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

next step is now  MNG-8258 activate Reproducible Builds by default
